### ISSUE TITLE
Question about CFG extraction.

### ISSUE THREAD

**Issue by Bowenduan** [radareorg/radare2#17674]
I want to extract the CFG(control flow graph) from a binary.  This can be done by using command `agf` at entry0. But the basic block defined by radare2 includes call instruction(like the screenshot below, there is a `call fcn.00467670` inside the basic block), which changes the execution order and should become the end of the basic block.
![image](https://user-images.githubusercontent.com/33278320/93566257-534f5800-f9bf-11ea-8ee4-102185f49c10.png)

  So, how to extract CFG where the basic block dosen't have call instruction inside ?

**Comment by ret2libc** [comment]
AFAIK it is common to consider the CFG of a function with calls as part of a basic block. I know there are cases where you want to have an even more granular CFG (I think Angr does that), but I think we don't have a way right now to directly get the CFG as you would like. Maybe we could add a `anal.XXXX` variable to control this behaviour in the future.

If you do need it now though, you can probably script a bit with r2pipe to split the provided basic blocks however you want. It is of course a bit more painful :+1:

@trufae am I missing any command?

**Comment by trufae** [comment]
in r1 there are some options to specify which instructions break blocks. so just tweaking that was enough to have a single CFG of the whole program at once.

in r2 we went formal and just split like everyone else do. i would suggest to improve the ab command to be able to list all the basic blocks, notjust the one in the current offset. using the table api and showing addr, true/fail branches, type of block, etc. you may be able to get the json out of it and process it like u want.

How does it sound to you? may this be a good solution to your question?

**Comment by Bowenduan** [comment]
So there is no command in radare2 to extract the CFG I explained, and the best way is to write a script myself. Is that right?

**Comment by ret2libc** [comment]
> So there is no command in radare2 to extract the CFG I explained, and the best way is to write a script myself. Is that right?

That is right. By the way, as far as I know you can't get the CFG as you want it even from IDA. On r2, however, you can implement it yourself if you want, differently from a closed source solution :)

**Comment by karliss** [comment]
One thing to note that in situations where you want calls to be considered a control graph edge there are hundred other cases that also need to be considered and depends on each specific use case, the assumptions you make, scripts and tools that will be consuming the results. Call instruction isn't only instruction that adds nontrivial control flow changes.
Some examples

* Do you need the actual edges to functions called or only splits at specified points to simplify  the script processing it depending on needs.
* What about syscalls? They can affect control flow both within program and across userspace/kernelspace. If you are analyzing kernel vulnerabilities you might be interested in combined control flow across program and kernelspace.
* How to deal with calls to imported functions? 
* How to deal with indirect calls and jumps? Do you model sets of potential jump targets, do you assume that anything can be reached in such situation? Maybe you analyze class inheritance and want edges either to all or one of the virtual functions that could be called in specific place. Maybe you are trying to describe system with control flow integrity mechanism. Maybe you don't care about such situations.
* What about `div` instruction instead of `call instruction`. It can cause controll flow changes through the interrupt handling mechanism. Do you want an edge to the interrupt handler within kernel or the registered handler within program? It's not unusual for obfuscated malware or commercial software to intentionally use such control flow for emulation or debugging detection or interfere with analysis. It might not even be obfuscated program but a programming language using such control flow for exception handling mechanism. In some programs exceptions are widely used, it may also be interesting target due to error handling code often being poorly tested and potentially vulnerable.
Why stop there do you also want a new basic block after each mov/ldr/str instruction those can also change control flow.
* Previous use-case too obscure? How about embeded software on microcontrollers? In embedded software interrupts can be part of normal control flow. No jumps across kernel/user space there. Those don't even have to be interrupts from external io, they can be intentionally software generated interrupts.
* Maybe you are using SMT solver and can prove that those indirect calls are actually fixed edges. What if it's a multi threaded software, is that proof still valid, can there be new edges? If you are trying prove correctness of an implementation for multi threading primitives  in a library or OS this might be important.
* What if you are trying to perform glitch analysis either for validating safety critical system or to extract a cryptographic secret. How does a control flow graph looks then? Can you skip a jump or maybe any instruction? Can you flip a bit in jump target?
Maybe you can make a jump load target from wrong register.
* What if you only want only calls where the the called function is used only once? In that case both the calling edge and return edge could be analyzed precisely. 

You can come up with many real world uses cases each with slightly different needs. In most of them there will be need to have at least some amount of code which prepares the CFG returned by r2 into exact format required by further analysis scripts. Those scripts might as well query data from r2 in multiple parts. Where do you draw the line.

Some things to considers:
* Just because r2 command doesn't have word graph in the name doesn't mean you can't look at the data as graph. Just because name of field in the structure returned by r2 doesn't say edge doesn't mean that isn't a graph. Just because different kind of edges are described using differently named fields doesn't mean that it isn't a graph. For example "list  refs in json" doesn't say graph but you can look it as graph. List of edges is one of the basic forms for describing graphs when teaching about them. Any table can be considered a graph, especially if some of columns describe relationship to other objects.
* Does r2 provides required information at all, in any form
* Does r2 provide required information in machine readable format
* How difficult is it to gather required information from r2 commands and how this effort compares to the effort compares to unavoidable glue code for using it in any kind of analysis script or existing tool. How does it compare to the difficulty of implementing the core logic of analysis script. I consider querying r2 commands either recursively or in a loop to gather required information to be the easy part compared to implementing any kind of interesting analysis on this data.
* How unique are the needs. Is it something that saves work for 5 out of 10 different kind of uses cases that you could do with it or is exact format expected for one specific script.
* NxM problem. Let's look at the same example with reference graph. It can be split into multiple dimensions:
    * type of reference: data, call, jump, ...
    * node type: instruction, basic block, function, class, source file, module, ...
    * output format: json, drawable dot, gml, ascii graphics, ...
  Should r2 provide `N * M * O` separate commands, one for each combination or few commands that allow querying list of references, which basic block, function, ... contains given address? And leave combining them possible in ways developers didn't expect to user scripts.

**Comment by karliss** [comment]
Back to the topic of this specific question it isn't the first time people have asked for something like this.

**Comment by XVilka** [comment]
Moreover, I think some of the answers to these questions could land in the book.

**Comment by trufae** [comment]
See my abl pr

**Comment by sh0ve1** [comment]
@Bowen-n 
hi, guy
Did you solve the problem? Now I have the same proble. So I want to know if you finished the script.

**Comment by Bowen-n** [comment]
> @Bowen-n hi, guy Did you solve the problem? Now I have the same problem. So I want to know if you finished the script.

Sorry, I didn't. CFGs with function calls inside are acceptable in my situation.

**Comment by sh0ve1** [comment]
@Bowen-n ok. Thank you for your reply.
Best.
